### DISCUSSION TITLE
Changing USB mode while still connected to PC for script development

### DISCUSSION THREAD

**Discussion by tekikai** [EdgeTX/edgetx#2159] [Ideas]
I just started developing LUA scripts for the Nirvana NV14 and noticed that I had to plug and unplug the radio every time I wanted to switch between modes (storage, joystick, etc.) or rather try out my new script. I. e.:

- write new script and try out in firmware simulator 
- plug in radio, select storage mode
- transfer files
- unplug radio
- test script (haptics in this case, just to test which duration of haptic feedback buzz would feel right)
- rinse and repeat

If I would like to change something bigger in the script, I have to replug the radio etc. - this will wear out the USB port of my computer / laptop and / or the USB port of the radio (whichever I use to reconnect for the file transfer) and I would love to have an option to switch back from USB storage mode to joystick mode to directly test my scripts without having to unplug / replug. 

Like some kind of "Reset USB line button" somewhere in the UI which would disconnect from the PC and re-pop the request which USB mode one wants. Or even reconnect USB when one changes the setting which USB mode should be default.

For now I have not found an option to do so and I find plugging / replugging over and over to test scripts to be a bit cumbersome (and as mentioned wearing out the ports should one write and test a lot of scripts).

One option would be to use a USB extension cable, keep it plugged in the computer, and just do the plugging / replugging at the extension cable end, so no essential hardware USB ports would get used out / wonky over time except for the extension cable and the cable from the radio (which can be easily replaced).

It would be great to have a way to transfer files and then to test the LUA scripts etc. directly on my radio (Nirvana NV14) without replugging USB. Maybe even sending files over Bluetooth?
The SD card on the Nirvana NV14 is hidden behind a screwed down post to hold the lanyard and plugging in and removing the SD card would wear out the port over time too so that is not an option.

Or am I overlooking something here? Or any ideas how you guys do this script testing process?

**Comment by gagarinlg** [comment]
I use the USB extension cord setup and I think that is what most of the devs do.
Another option would be using a USB hub with switches for the ports.
Removing the SD card while the radio is on is a bad idea.
The Bluetooth idea sounds nice, I am just not sure if the Bluetooth module is able to do this with a uart connection between the module and the MCU.

**Reply by tekikai** [reply]
Thanks for the quick reply. I also figured that for now an extension cord is probably in order :).
Also I don't know if it is even possible in software / firmware to 'reset the USB line' so that the requester pops up again - so maybe not even possible in the future.
USB hub with switches sounds practical too!

**Reply by pfeerick** [reply]
I use a USB hub with power switches on it, so can 1) leave multiple leads plugged in and 2) just turn on and off as needed. A lot of the testing for Lua can be done in the Companion simulator, but at the end of the day you can't get around having to test on the radio. Having softare disconnect/reset capability would be nice, and worth looking into, but the path of least resistance is a hub with switches or usb extension for now.

**Reply by tekikai** [reply]
Ordered a switched hub now and cable (needed a new hub anyway) and will try what I like best (I think the hub 🤔).
And yes, testing haptics for example is best done on the radio I guess. 

Am crosspurposing my Nirvana NV14 as a controller for Elite Dangerous at the moment and learning more about 'special functions', 'logic switches' and some LUA at the same time (I wanted to have a custom haptic feedback on throttle center for Elite as there is no physical detent at the throttle mid-point).

**Reply by pfeerick** [reply]
Yes. And your point over in the haptic issue is valid, it's really only needed for the simulator. Elite Dangerous... interesting... never thought to use a TX as a interface for that... might have to give that a try :) You should be able to do that easily enough with LS and SF... LS that looks for the throttle position, and then SF that Plays Haptic if that LS is true. Although we would really need that haptic pattern thing that was talked about before as not on the stock haptics really suit... just want a normal short buzz.

**Reply by banyaszg** [reply]
I know this is an old discussion, but I implemented a similar thing during the implementation of my USB Joystick extension. It is not exactly this feature, but similar. When I want to change the USB Joystick interface on the fly, then I do a software USB disconnect/reconnect. @pfeerick If there is a demand for this kind of feature, then I think I can make it.

**Reply by pfeerick** [reply]
That sounds like a good idea. I'm not sure how it would be exposed when the device is in storage mode, as there is no UI visible at that point. However, that screen can be revisited, as we wanted to change the appearance of it anyway.

**Reply by banyaszg** [reply]
Maybe a global function? I think mainly the developers would use this feature, and as a global function they could bind it to a switch. However I didn't really checked the global functions yet.

**Reply by pfeerick** [reply]
Hm... yeah, that would be a good way to do it... GF/SF if triggered/active would do the disconnect... I like it! :)

**Reply by banyaszg** [reply]
Then this will be the way. I will work on it after I finish the Joystick interface.

**Reply by Eldenroot** [reply]
@banyaszg - cool!

**Reply by gagarinlg** [reply]
I do not think global functions are run when the radio is in USB storage mode. The main loop is short circuited to check the USB connectivity and that's it, if I remember correctly

**Reply by banyaszg** [reply]
Thanks! I'll check that part what the options are.
